### PR TITLE
Stretchable icons

### PR DESCRIPTION
**Pull request by kkaefer**
Fixes https://github.com/mapbox/mapbox-gl-native/issues/16017

Implements stretchable icons, porting from JS.

### CODE REVIEW

File: src/mbgl/layout/symbol_layout.cpp [hunk 1]
```diff
@@ -928,6 +961,19 @@ size_t SymbolLayout::addSymbol(SymbolBucket::Buffer& buffer,
     return index;
 }
 
+size_t SymbolLayout::addSymbols(SymbolBucket::Buffer& buffer,
+                                const Range<float> sizeData,
+                                const SymbolQuads& symbols,
+                                const Anchor& labelAnchor,
+                                PlacedSymbol& placedSymbol,
+                                float sortKey) {
+    size_t lastIndex = -1;
```
Comment (alexshalamov) [inline src/mbgl/layout/symbol_layout.cpp]: 0u? -1 would be max size_t

**Comment by kkaefer** [comment]
Remaining questions:

- We're using `std::vector` for storing stretches, which might be using quite a bit of memory. Ideally, we'd use a "small vector" that stores 0-1 elements inline to avoid heap allocations.
- We're storing texture coordinates as integers in Native. I believe we do that as well in JS, which means that we'll have to remove the option to pass floating point numbers in stretches.

**Comment by ansis** [comment]
> We're storing texture coordinates as integers in Native. I believe we do that as well in JS, which means that we'll have to remove the option to pass floating point numbers in stretches.

Yep, should we do something to enable some fractional values? I could see half pixel values being useful but not sure about anything beyond that.

**Comment by ansis** [comment]
I took a look a look at the core logic part and I checked that all the followup fixes I made in -js are also included here. Those look good (assuming the TODOs are done). I'll leave the rest of the review to Alex

File: platform/android/src/map/image.cpp [hunk 2]
```diff
@@ -26,7 +26,7 @@ mbgl::style::Image Image::getImage(jni::JNIEnv& env, const jni::Object<Image>& i
 
     mbgl::PremultipliedImage premultipliedImage({ static_cast<uint32_t>(width), static_cast<uint32_t>(height) });
     if (premultipliedImage.bytes() != uint32_t(size)) {
-        throw mbgl::util::SpriteImageException("Sprite image pixel count mismatch");
```
[the remark this answers is not in the document]
Comment (alexshalamov) [reply platform/android/src/map/image.cpp]: Not related to this PR, should be fixed separately.

**Comment by kkaefer** [comment]
@alexshalamov thanks for rebaselining!

We're using a lot of `std::vector`s that have no or only one element. my fear is that std::vector is using too much memory + heap allocations. Is there a tutorial for how I could benchmark this in GL native? I'm looking at https://github.com/abseil/abseil-cpp/blob/master/absl/container/inlined_vector.h; have we considered including abseil (or some other library) that would give us these types?

**Comment by alexshalamov** [comment]
@kkaefer I created an issue https://github.com/mapbox/mapbox-gl-native/issues/16118 to follow-up idea of using inlined vectors.

**Pull request merged**